src/core/fair_queue.cc

```diff
@@ -159,6 +159,14 @@ class fair_queue::priority_class_data {
     void update_shares(uint32_t shares) noexcept {
         _shares = (std::max(shares, 1u));
     }
+
+    bool has_noncancelled_entry() const {
+        bool result = false;
+        for (auto& ent : _queue) {
+            result |= ent._capacity > 0;
+        }
+        return result;
```

Comment:
```suggestion
        auto found = boost::range::find_if(_queue, [] (const auto& ent) {
            return ent._capacity > 0;
        });
        return found != _queue.end();
```
no need to continue if one of the non-cancelled entry is found.

Reply:
That was my initial version but it makes error case faster by making correct/happy/usual path slower by introducing branching inside the loop. I think it's better to have faster happy path instead of faster error path which ends in assertion crashing anyway.

Reply:
If we aim at speed and think that there can be canceled requests in the queue, then it's worth checking `io_queue::_queued_requests` to become 0 in `~io_queue()` and drop the queue being empty part of this assertion